dynamiqs/_checks.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+import jax.numpy as jnp
+from jax import Array
+
+_is_perfect_square = lambda n: int(n**0.5) ** 2 == n
+
+
+def has_shape(x: Array, shape: str) -> bool:
+    cases = {
+        '(..., n, 1)': lambda x: x.ndim >= 2 and x.shape[-1] == 1,
+        '(..., 1, n)': lambda x: x.ndim >= 2 and x.shape[-2] == 1,
+        '(..., n, n)': lambda x: x.ndim >= 2 and x.shape[-2] == x.shape[-1],
+        '(N, ..., n, n)': lambda x: x.ndim >= 3 and x.shape[-2] == x.shape[-1],
+        '(..., n)': lambda x: x.ndim >= 1,
+        '(n, 1)': lambda x: x.ndim == 2 and x.shape[-1] == 1,
+        '(1, n)': lambda x: x.ndim == 2 and x.shape[-2] == 1,
+        '(n, n)': lambda x: x.ndim == 2 and x.shape[-2] == x.shape[-1],
+        '(n,)': lambda x: x.ndim == 1,
+        '(N, n, 1)': lambda x: x.ndim == 3 and x.shape[-1] == 1,
+        '(N, n, n)': lambda x: x.ndim == 3 and x.shape[-2] == x.shape[-1],
+        '(?, n, n)': lambda x: 2 <= x.ndim <= 3 and x.shape[-2] == x.shape[-1],
+        '(..., n^2, 1)': lambda x: x.ndim >= 2
+        and _is_perfect_square(x.shape[-2])
+        and x.shape[-1] == 1,
+    }
+    if shape in cases:
+        return cases[shape](x)
+    else:
+        raise ValueError(f'Unknown shape specification `{shape}`.')
+
+
+def check_shape(
+    x: Array, argname: str, *shapes: str, subs: dict[str, str] | None = None
+):
+    for shape in shapes:
+        if has_shape(x, shape):
+            return
+
+    if len(shapes) == 1:
+        shapes_str = shapes[0]
+    else:
+        shapes_str = ', '.join(shapes[:-1]) + ' or ' + shapes[-1]
+
+    if subs is not None:
+        for k, v in subs.items():
+            shapes_str = shapes_str.replace(k, v)
```

Comment:
I didn't catch what this `subs` does. It's substitions for the shape?

Reply:
This is needed because we have a specification for shapes which does not work for arbitrary symbols, but sometimes the doc uses more complex symbols, and the error message should as well. For example in `dq.expect()`:
```python
# O _(array_like of shape (nO?, n, n))_: Arbitrary operator or list of _nO_ operators.
check_shape(O, 'O', '(?, n, n)', subs={'?': 'nO?'})
```
This says « verify the shape of O with specification '(?, n, n)', but in the error message replace '?' by 'nO?' ».

It's just used in `dq.expect()` for now, but will be used a lot by `sesolve`/`mesolve`, e.g.:
```python
# rho0 (array-like of shape (nrho0?, n, 1) or (nrho0?, n, n)) – Initial state.
check_shape(rho0, 'rho0', '(?, n, 1)', '(?, n, n)', subs={'?': 'nrho0?'})
```

I'll add a comment to make this more clear.